fastlane/Fastfile

```diff
@@ -879,54 +878,83 @@ def simulator_version
-# expected to look like a series of Fs and Ts.
+# Test accounts info is stored in JSON-formatted string, where keys represent
+# the account index, and values are either "free" or CIRCLE_BUILD_NUM, in case if account
+# was occupied by a buld. Returns the hash.
 # Example:
-#   get_test_accounts_state
-#     => "TFFFF"
-def get_test_accounts_state
-  uri = URI.parse(get_required_env("UI_TESTS_ACCOUNTS_URL"))
+#   get_test_accounts_hash
+#     => {"0"=>"4079", "1"=>"free", "2"=>"free", "3"=>"free"}
+def get_test_accounts_hash
+  uri = URI.parse(get_required_env("UI_TESTS_ACCOUNTS_JSON_URL"))
   response = Net::HTTP.get_response(uri)
   accounts_state = response.body.chomp
-  UI.message("State of test accounts: #{accounts_state}")
-  accounts_state
+  JSON.parse(accounts_state)
 end
 
-# Returns the index of first "F" (Free) occurence in the test
-# accounts keyvalue, or crashes fastlane if it was not found.
-# Example:
-#   find_free_test_account
-#     => 2
+# Finds the index of first free account in accounts hash, and marks it as accupied by current build.
+# Crashed fastlane if free account was not found.
 def find_free_test_account
-  $used_test_account_index = get_test_accounts_state.index("F")
+  accounts_hash = get_test_accounts_hash
+  UI.message("Looking for a free test account...")
+  UI.message("Accounts state: #{accounts_hash}")
+
+  accounts_hash.each do |key, value|
+    break $used_test_account_index = key.to_i if value == "free"
+  end
+
   UI.user_error!("Could not find free UI Test account. Quitting.") if $used_test_account_index.nil? 
   UI.message("Free account index: #{$used_test_account_index}")
+  change_test_account_availability("#{get_required_env("CIRCLE_BUILD_NUM")}")
 end
 
-# Replaces a character at `index` with `availability` value.
+# Replace a value in a key which is equal to $used_test_account_index global variable
 # Example:
-#   change_test_account_availability("F")
-#     => FFFFF
+#   change_test_account_availability("free")
+#     => {"0"=>"free", "1"=>"free", "2"=>"free", "3"=>"free"}
+#
+#   change_test_account_availability("4820")
+#     => {"0"=>"4820", "1"=>"free", "2"=>"free", "3"=>"free"}
 def change_test_account_availability(availability)
-  accounts_state = get_test_accounts_state
-  accounts_state[$used_test_account_index] = availability
-  UI.message("Writing new account state: #{accounts_state}")
-  uri = URI.parse(get_required_env("UI_TESTS_ACCOUNTS_URL"))
+  accounts_hash = get_test_accounts_hash
+  accounts_hash[$used_test_account_index.to_s] = availability
+  submit_accounts_hash(accounts_hash)
+end
+
+def sanitize_test_accounts
+  accounts_hash = get_test_accounts_hash
+  UI.message("Sanitizing stale test accounts...")
+  UI.message("Accounts before sanitizing: #{accounts_hash}")
+
+  accounts_hash.each do |key, value|
+    if value != "free" && is_job_running(value) == false
+      accounts_hash[key] = "free"
+      submit_accounts_hash(accounts_hash)
+    end
+  end
```

Comment:
You seem to post (submit) the JSON on each iteration in which you find a case to sanitize, is that on purpose, to limit the race condition given `is_job_running` requires a network call and can thus take time? (As opposed to only submitting the updated hash at the end)

```suggestion
  accounts_hash.each do |key, value|
    next if value == "free" || is_job_running(value)
    accounts_hash[key] = "free"
    submit_accounts_hash(accounts_hash)
  end
```

Reply:
This is a beautiful improvement of readability, thanks @AliSoftware (and thanks for jumping in, in general)!

> You seem to post (submit) the JSON on each iteration in which you find a case to sanitize, is that on purpose, to limit the race condition given is_job_running requires a network call and can thus take time? (As opposed to only submitting the updated hash at the end)

I was thinking of limiting the race condition initially, yes. Now, after trying submitting the hash only once at the end, I realize that another reason for doing this the initial way is that we are sure we submit changes only when needed. If no changes were made to the hash, but we move to submit to the end, we will need to additionally make sure that we only submit if the final hash is different from the initial, i.e. `submit_accounts_hash(final_accounts_hash) if final_accounts_hash != accounts_hash`. And this is actually what I did 🙂 

I'm really not sure what is better - having less code/variables or having less requests? 🤔

Reply:
That's a good question. I think updating on every loop rather than at the end will still be safer tho.

Especially imagine job 1 starts and do the cleanup only after checking the status of _all_ jobs occupying an account, which could take a couple of seconds in total. While it does this, job 2 starts and start doing the same.

After the few seconds that was required by job 1 to do all the requests to check the status of all jobs occupying an account, it finally uploads the updated JSON, picks an account for itself, and upload again.

In parallel job 2 will still be processing the JSON from before job 1 updated anything and will not see that job 1 has since taken one of the accounts for itself…